### ISSUE TITLE
Wan 2.2

### ISSUE THREAD

**Issue by mstojkovicTT** [tenstorrent/tt-xla#4291]
## Summary

Wan 2.2 TI2V-5B is a unified 5B-parameter diffusion transformer from Alibaba's Wan team that generates videos from text prompts (T2V) or from a reference image + text prompt (I2V). Both modes share the same checkpoint. I2V reuses the same channel layout but conditions via a per-frame blend + per-patch timestep mechanism.

**Resolution plan**: start with **480p**, then scale to **720p**.

## Work Items
- [ ] 480p T2V — Multi-Device Sharded (LoudBox / Galaxy)
- [ ] 480p I2V — Multi-Device Sharded (LoudBox / Galaxy)
- [ ] 720p T2V — Multi-Device Sharded (LoudBox / Galaxy)
- [ ] 720p I2V — Multi-Device Sharded (LoudBox / Galaxy)
- [ ] Try 14B variant with the same goal

---

## Architecture Overview

The pipeline has 4 components: **UMT5-XXL Text Encoder**, **3D Causal VAE Encoder** (I2V only), **WanDiT Transformer** (the denoiser), and **3D Causal VAE Decoder**.

```
    "A cat walking on snow"                        input_image.jpg  (I2V only)
            │                                            │
            ▼                                            ▼
  ┌───────────────────┐                      ┌───────────────────┐
  │   UMT5-XXL        │                      │  3D Causal VAE    │
  │   Text Encoder    │                      │  Encoder          │
  │                   │                      │                   │
  │  IN:  tokens      │                      │  IN:  (1,3,1,H,W) │
  │       (1, 512)    │                      │  OUT: (1,48,1,    │
  │  OUT: (1,512,     │                      │        H/16,W/16) │
  │        4096)      │                      │                   │
  └────────┬──────────┘                      └────────┬──────────┘
           │                                          │
           │ cross-attention                          │ becomes "condition"
           │                                          │ (padded temporally)
           │                                          ▼
           │                              ┌──────────────────────┐
           │                              │  Latent Assembly     │
           │                              │                      │
           │                              │  noise:      (48,T',H',W')
           │                              │  condition:  (48,T',H',W')
           │                              │  first_frame (1, T',H',W')
           │                              │    _mask:                 │
           │                              │                      │
           │                              │  At each step:       │
           │                              │    input = (1-mask)*condition
           │                              │          + mask*noise
           │                              │  timestep = mask * t │
           │                              └────────┬─────────────┘
           │                                       │
           ▼                                       ▼
  ┌─────────────────────────────────────────────────────────────┐
  │                   WanDiT Transformer (5B)                   │
  │                                                             │
  │  30 layers · dim 3072 · 24 heads × 128 · FFN 14336          │
  │  3D Full Self-Attention · Cross-Attention · AdaLN-Zero      │
  │  3D RoPE · Flow Matching (v-prediction)                     │
  │                                                             │
  │  IN:  hidden_states      (1, 48, T', H', W')                │
  │       encoder_hidden_st. (1, 512, 4096)                     │
  │       timestep           (1, num_patches) per-patch         │
  │  OUT: velocity           (1, 48, T', H', W')                │
  │                                                             │
  │  Denoising loop: 50 Euler steps, CFG scale 5.0              │
  │  Per step: 2 forward passes (cond + uncond) = 100 total     │
  └──────────────────────────┬──────────────────────────────────┘
                             │
                             ▼
                   ┌───────────────────┐
                   │  3D Causal VAE    │
                   │  Decoder          │
                   │                   │
                   │  IN:  (1,48,T',   │
                   │        H',W')     │
                   │  OUT: (1,3,T,H,W) │
                   └────────┬──────────┘
                            │
                            ▼
                    MP4 @ 24fps
```

### T2V vs I2V — Same Channels, Different Conditioning

The TI2V-5B DiT takes **48-channel latents** in both modes (unlike Wan 2.1 I2V which uses channel concatenation). Conditioning is done via a **blend** at each denoising step, plus **per-patch timesteps**.

**T2V mode**: The VAE Encoder is **not used**. `condition` is all zeros, `first_frame_mask` is all ones. The blend formula reduces to pure noise:
```python
latent_model_input = (1 - 1) * 0 + 1 * noise = noise
timestep = 1 * t = t  (uniform)
```

**I2V mode** (`expand_timesteps=True`): The VAE Encoder runs once on the reference image. At each step, the first frame is replaced with the VAE-encoded image and its timestep is forced to 0 (treated as clean):
```python
# first_frame_mask: 0 at frame 0, 1 elsewhere
latent_model_input = (1 - mask) * condition + mask * noise
#                    = condition at frame 0, noise at frames 1..T-1

timestep = mask * t
#          = 0 at frame 0 (model treats as already clean)
#          = t at frames 1..T-1 (normal denoising)
```

This forces the model to preserve the input image while generating plausible subsequent frames.

---

## Shape Calculation

**Formula** (VAE stride `(4, 16, 16)`, DiT patch size `(1, 2, 2)`):

| Latent Dim | Formula |
|---|---|
| `T'` (temporal latent frames) | `(T_video − 1) // 4 + 1` |
| `H'` (height latent) | `H_pixel // 16` |
| `W'` (width latent) | `W_pixel // 16` |
| Patches per step | `T' × (H' // 2) × (W' // 2)` |

Pixel dims are rounded down to multiples of `16 × 2 = 32` before VAE encode (pipeline-enforced, ensures clean divisibility).

### 480p (initial target)

- Video resolution: **480 × 832**
- Frames: **81** (~3.4s @ 24fps)

| Value | Calculation | Result |
|---|---|---|
| `T'` temporal latent | (81 − 1) / 4 + 1 | **21** |
| `H'` height latent | 480 / 16 | **30** |
| `W'` width latent | 832 / 16 | **52** |
| DiT input shape | (1, 48, 21, 30, 52) | |
| DiT output shape | (1, 48, 21, 30, 52) | |
| Patches per step | 21 × 15 × 26 | **8,190** |
| Total DiT calls | 50 steps × 2 (CFG) | **100** |

### 720p (scale-up target)

Input size `720 × 1280` is rounded down to **`704 × 1280`** by the pipeline (multiple of 32).

- Video resolution: **704 × 1280** (post-alignment)
- Frames: **121** (~5s @ 24fps)

| Value | Calculation | Result |
|---|---|---|
| `T'` temporal latent | (121 − 1) / 4 + 1 | **31** |
| `H'` height latent | 704 / 16 | **44** |
| `W'` width latent | 1280 / 16 | **80** |
| DiT input shape | (1, 48, 31, 44, 80) | |
| DiT output shape | (1, 48, 31, 44, 80) | |
| Patches per step | 31 × 22 × 40 | **27,280** |
| Total DiT calls | 50 steps × 2 (CFG) | **100** |

### Why this matters

**720p has 3.3× more patches than 480p** (27,280 vs 8,190). Self-attention is quadratic in sequence length, so **720p attention is ~11.1× more compute** per forward pass than 480p. That's why we start with 480p on a single device, then use multi-device sharding to make 720p feasible.

### Fixed Shapes (both resolutions)

| Tensor | Shape | Notes |
|---|---|---|
| Text tokens | `(1, 512)` | UMT5 tokenizer, max sequence length |
| Text embeddings | `(1, 512, 4096)` | UMT5-XXL hidden dim |
| DiT channels in/out | **48** | same in both T2V and I2V mode |
| Model params | 5B | 30 blocks, dim 3072, 24 heads |

---

### 1. UMT5-XXL Text Encoder
**Status**: TBD

Frozen text encoder (~4.7B params). Only runs once per generation (not per denoising step) so it's low priority for device bring-up.

- IN: `input_ids` (1, 512), `attention_mask` (1, 512)
- OUT: `last_hidden_state` (1, 512, 4096)

### 2. 3D Causal VAE Encoder (I2V only)
**Status**: TBD

Encodes the reference image into latent space. Single forward pass per generation. Causal convolutions ensure each frame depends only on current/past frames.

Config: `base_dim=160`, `z_dim=48`, `dim_mult=[1,2,4,4]`, `num_res_blocks=2`, internal `patch_size=2` (groups 2×2 pixels into 12 channels before encoder), total spatial stride **16**, temporal stride **4**.

| Resolution | Input | Output |
|---|---|---|
| 480p | (1, 3, 1, 480, 832) | (1, 48, 1, 30, 52) |
| 720p | (1, 3, 1, 704, 1280) | (1, 48, 1, 44, 80) |

**Not used in T2V mode.**

### 3. WanDiT Transformer (5B) 
**Status**: TBD

**Architecture** (from HuggingFace config):
- **30 transformer blocks**, hidden dim **3072** (24 heads × 128 head_dim)
- FFN dim **14336** with SwiGLU activation
- 3D RoPE positional embeddings (temporal + 2D spatial), `rope_max_seq_len=1024`
- AdaLN-Zero modulation driven by timestep embedding
- Flow Matching (rectified flow) with v-prediction
- `in_channels=48`, `out_channels=48`, `text_dim=4096`
- `image_dim=None`, `added_kv_proj_dim=None` -> no CLIP image encoder (unlike Wan 2.1)

**Sequence length per pass**:
- 480p: **8,190 patches**
- 720p: **27,280 patches**

### 4. 3D Causal VAE Decoder
**Status**: TBD

Decodes denoised latent back to pixel space. Single forward pass per generation. Expected to hit OOM on device for full resolution due to upsampling activations.

| Resolution | Input | Output |
|---|---|---|
| 480p | (1, 48, 21, 30, 52) | (1, 3, 81, 480, 832) |
| 720p | (1, 48, 31, 44, 80) | (1, 3, 121, 704, 1280) |

---

### What's Done
- 

### 480p (81 frames, 480×832)
```
Input image (I2V):   (1, 3, H_orig, W_orig)  →  Resize  →  (1, 3, 480, 832)
Text prompt:         string  →  UMT5 tokenize  →  (1, 512)
                            →  UMT5 encode    →  (1, 512, 4096)

VAE Encode (I2V):    (1, 3, 1, 480, 832)  →  (1, 48, 1, 30, 52)
Gaussian Noise:      sample N(0,1)        →  (1, 48, 21, 30, 52)
Condition (I2V):     image latent padded  →  (1, 48, 21, 30, 52)
First frame mask:    0 at frame 0, 1 else →  (1, 1, 21, 30, 52)

Per step (I2V):
  model_input = (1 - mask) * condition + mask * noise   (48, 21, 30, 52)
  timestep    = (mask[:,:,::2,::2] * t).flatten()       (num_patches=8190,)

DiT Denoising:       (1, 48, 21, 30, 52) + text_emb (1, 512, 4096)
                       →  50 Euler steps (× 2 forward passes with CFG)
                       →  (1, 48, 21, 30, 52)

VAE Decode:          (1, 48, 21, 30, 52)  →  (1, 3, 81, 480, 832)
Postprocess:         (1, 3, 81, 480, 832)  →  MP4 @ 24fps
```

### 720p (121 frames, 704×1280 after alignment)
```
Input image (I2V):   (1, 3, H_orig, W_orig)  →  Resize  →  (1, 3, 704, 1280)
Text prompt:         string  →  UMT5 tokenize  →  (1, 512)
                            →  UMT5 encode    →  (1, 512, 4096)

VAE Encode (I2V):    (1, 3, 1, 704, 1280)  →  (1, 48, 1, 44, 80)
Gaussian Noise:      sample N(0,1)         →  (1, 48, 31, 44, 80)
Condition (I2V):     image latent padded   →  (1, 48, 31, 44, 80)
First frame mask:    0 at frame 0, 1 else  →  (1, 1, 31, 44, 80)

Per step (I2V):
  model_input = (1 - mask) * condition + mask * noise   (48, 31, 44, 80)
  timestep    = (mask[:,:,::2,::2] * t).flatten()       (num_patches=27280,)

DiT Denoising:       (1, 48, 31, 44, 80) + text_emb (1, 512, 4096)
                       →  50 Euler steps (× 2 forward passes with CFG)
                       →  (1, 48, 31, 44, 80)

VAE Decode:          (1, 48, 31, 44, 80)  →  (1, 3, 121, 704, 1280)
Postprocess:         (1, 3, 121, 704, 1280)  →  MP4 @ 24fps
```

---

## References

- **Model**: [Wan-AI/Wan2.2-TI2V-5B-Diffusers](https://huggingface.co/Wan-AI/Wan2.2-TI2V-5B-Diffusers)
- **Repo**: https://github.com/Wan-Video/Wan2.2
- **Architecture**: diffusers `WanTransformer3DModel` + `AutoencoderKLWan` + `UniPCMultistepScheduler`
